### PR TITLE
React / Sutil inspired state handling

### PR DESCRIPTION
**Pull request by JaggerJo**
This is a prototype for a React/Sutil inspired state handling library for FuncUI. 

There is an example contact app included to test it. 

Screenshot 2021-12-13 at 00 23 49

Feedback is appreciated.

### CODE REVIEW

File: src/Avalonia.FuncUI.Components/Wires.fs [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    abstract InstanceId: Guid with get
+
+type ITap<'signal> =
+    inherit IConnectable
+    abstract member CurrentSignal: 'signal with get
+    abstract member Subscribe : ('signal -> unit) -> IDisposable
+
+type IWire<'signal> =
+    inherit ITap<'signal>
+    abstract member Send : 'signal -> unit
```
Comment (JaggerJo) [inline src/Avalonia.FuncUI.Components/Wires.fs:16]: Not really happy with the naming here. Open for suggestions.

File: src/Examples/Examples.ContactBook/Views.fs [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+
+open System
+open Avalonia.Controls
+open Avalonia.FuncUI.Components
+open Avalonia.FuncUI.DSL
+open Avalonia.FuncUI.Types
+open Avalonia.Layout
+open Avalonia.FuncUI
+
+let contactListView (contacts: ITap<Contact list>, selectedId: IWire<Guid option>, filter: IWire<string option>) =
```
Comment (JaggerJo) [inline src/Examples/Examples.ContactBook/Views.fs:11]: example usage

**Comment by JordanMarr** [comment]
This looks really cool!

**Comment by JordanMarr** [comment]
Looking at your ContactBook example now... 
Your componentized app is very clean. I love it! So happy to see this.

**Comment by JaggerJo** [comment]
> Looking at your ContactBook example now...
> Your componentized app is very clean. I love it! So happy to see this.

Great to hear!

There are a few things I'm not yet sure about:

1. Naming. `IWire` is terrible, not sure if `IReadable` + `IWritable` or a variation of it would be better. 

2. When using a `useState` hook in react its value is consistent during one render. 

```fsharp
let counter = React.useState 0
React.useEffectOnce (fun _ -> 
    counter.set (counter.current + 1)
    counter.set (counter.current + 1)
    counter.set (counter.current + 1)
)
counter.set (counter.current + 1)
counter.set (counter.current + 1)
counter.set (counter.current + 1)

// current render 'counter' = 0
// next render 'counter' = 1
```

**Comment by JordanMarr** [comment]
> 1. Naming. `IWire` is terrible, not sure if `IReadable` + `IWritable` or a variation of it would be better.

I like publish/subscribe because it is used a lot in UI frameworks for sharing between view models in MVVM. 

Rx uses Observable/Observer/Subject (where Subject can publish and subscribe). Still not as clear as publish/subscriber though IMO.

> 2. When using a `useState` hook in react its value is consistent during one render.
> 
> ```fsharp
> let counter = React.useState 0
> React.useEffectOnce (fun _ -> 
>     counter.set (counter.current + 1)
>     counter.set (counter.current + 1)
>     counter.set (counter.current + 1)
> )
> counter.set (counter.current + 1)
> counter.set (counter.current + 1)
> counter.set (counter.current + 1)
> 
> // current render 'counter' = 0
> // next render 'counter' = 1
> ```
> 
> Not sure yet if we should copy this behaviour.

I don’t have a strong opinion because I am wary of multiple renders in React and so I would never do this - I would just set one state at time. I also only use a single useState at a time to prevent confusing multi pass render scenarios. 

> 
> 4. We currently depend on call order or hooks like react. I don't really like this as we can't conditionally add hooks to the component.
> 
> Alternatives to make a call to ctx.useWhatever unique could be:
> 
> * implicitly passing a key to `ctx.useWhatever` that we supply via the [CallerLineNumberAttribute](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.callerlinenumberattribute?view=net-6.0)
> * explicitly passing a key

I am already used to React not allowing conditional hooks, so this limitation doesn’t really bother me.

**Comment by Zaid-Ajaj** [comment]
One of the most exciting features I have seen in a while for F# on desktop 😍 looking forward to dig into this further. 

One thing I didn't see in the example is the array dependencies for effects which re-triggers the effect when one of it's dependencies change:
```fs
// currently this is the API usage
let contact = ctx.useTap contact
let image = ctx.useAsync Api.randomImage

// In case the image is _dependant_ on the contract
let contact = ctx.useTap contact
let image = ctx.useAsync(Api.contactImage contact.Id, [ contact.Id ])
```

**Review by JordanMarr — APPROVED** [review]
The new names are much more representative of their intended functionality in this context. Very nice!

File: src/Examples/Examples.ContactBook/Views.fs [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+        | None ->
+            TextBlock.create [
+                DockPanel.dock Dock.Top
+                TextBlock.dock Dock.Right
+                TextBlock.text "-"
+            ] :> IView
+    )
+
+let mainView () =
+    Component (fun ctx ->
```
[the remark this answers is not in the document]
Comment (JaggerJo) [reply src/Examples/Examples.ContactBook/Views.fs:267]: The constructor returns a `Component` (inherits `ContentControl`) while the `create` function returns a `View<Component>` that can be used in the DSL context. 

So one if for the DSL context and one if for Avalonia context.

**Comment by JaggerJo** [comment]
@JordanMarr 

> I wonder if there would be value in utilizing the built-in IObservable<'T>, IObserver<'T> and Subject<'T> for the underlying pub/sub mechanism.
While I admittedly don't quite have my head wrapped around the nuances of everything Wire.fs, it seems like it is very similar.
>
> Potential benefits might be:
> 
> Mainly, the ability for an end user to take advantage of more advanced Rx capabilities on the incoming subscriptions.
For example: if you wanted to share an IObservable<string> of keystrokes from one component to another, the user could use Rx debounce to limit observed messages to avoid unnecessary refreshes. This is a fairly common request that would be a pain to implement manually.
A secondary benefit would be a familiar pub/sub model with known names (IObservable instead of ITap, Subject instead of IWire, etc).

Initially considered this, but `IObservable` does not contain a value. 

`IValue` could/should implement `IObservable` I think, but It's not sufficient on it's own. 

https://fsprojects.github.io/FSharp.Data.Adaptive/ could work instead of our own values, but I don't fully get it. From looking at the code it does a lot I don't understand.

File: src/Avalonia.FuncUI.Components/Wires.fs (hunk 1 above)
Comment (JordanMarr) [reply src/Avalonia.FuncUI.Components/Wires.fs]: > Initially considered this, but `IObservable` does not contain a value.

I had to look it up, but `IEvent` already inherits from `IObservable`, so it seems that the door is still open to allowing integration with Rx.

**Comment by JaggerJo** [comment]
> One of the most exciting features I have seen in a while for F# on desktop 😍 looking forward to dig into this further.
> 
> One thing I didn't see in the example is the array dependencies for effects which re-triggers the effect when one of it's dependencies change:
> 
> ```fsharp
> // currently this is the API usage
> let contact = ctx.useTap contact
> let image = ctx.useAsync Api.randomImage
> 
> // In case the image is _dependant_ on the contract
> let contact = ctx.useTap contact
> let image = ctx.useAsync(Api.contactImage contact.Id, [ contact.Id ])
> ```

@Zaid-Ajaj Thanks for the kind words. I have time next week to hopefully make some progress on this. If you have any suggestions just let me know. Happy to discuss anything.

**Comment by sleepyfran** [comment]
I haven't had a chance to properly check all of the code, but I gotta say that this looks absolutely amazing and really promising! 😃 

Just a couple of thoughts I have in my head after a quick check:
- After your latest commit I have to say that the naming is much clearer now, but I feel like the hooks could be more self-explanatory like: `useValue` -> `useState`; `usePassedValue` -> `usePassedState` or `useExternalState`. Basically to better define the notion that a value can be modified and acts as a state holder. Wouldn't say it's a must since the current naming is not bad at all, but it took me a second to realize that `useValue` meant creating a state.
- I was going over the list of built-in React hooks and I think making our own version of `useEffect` (I'd say even two: `useEffect` that takes a dependency array and a `useOnStartEffect` or `useEffectOnce` that executes only the first time) and `useMemo` would be a great addition to the current bunch.
- `useAsync` is an **awesome** idea and makes such a common thing incredibly easy to implement, awesome job!

**Comment by JaggerJo** [comment]
> I haven't had a chance to properly check all of the code, but I gotta say that this looks absolutely amazing and really promising! 😃 

Still an early prototype. 

> Just a couple of thoughts I have in my head after a quick check:
> 
> - After your latest commit I have to say that the naming is much clearer now, but I feel like the hooks could be more self-explanatory like: `useValue` -> `useState`; `usePassedValue` -> `usePassedState` or `useExternalState`. Basically to better define the notion that a value can be modified and acts as a state holder. Wouldn't say it's a must since the current naming is not bad at all, but it took me a second to realize that `useValue` meant creating a state.

Like your suggestions, would favor names that are different from react. I fear (and I might be wrong) that having similar names with different behavior could confuse people.

> - I was going over the list of built-in React hooks and I think making our own version of `useEffect` (I'd say even two: `useEffect` that takes a dependency array and a `useOnStartEffect` or `useEffectOnce` that executes only the first time) and `useMemo` would be a great addition to the current bunch.

I have a implementation that implements effects. It currently is super buggy, will post here in a few days with more details/ problems that come with effects.

> - `useAsync` is an **awesome** idea and makes such a common thing incredibly easy to implement, awesome job!

**Comment by JaggerJo** [comment]
## Update

### Better Names
```fsharp
Component (fun ctx ->
    let contacts = ctx.usePassedState ContactStore.shared.Contacts
    let selectedId = ctx.useState (None: Guid option)
    let filter = ctx.useState (None: string option)
    ..
)
```

### Conditionally render hooks. 
This is archived by using the line number a hook was invoked from as the identity instead of the call order. The line number is implicitly passed. 

```fsharp
type Context with
    member this.useState<'value>(init: 'value, [<CallerLineNumber>] ?identity: int) : IWritable<'value> =
        this.useHook<'value>(
            StateHook.Create (
                identity = $"Line: {identity.Value}",
                state = StateHookValue.Lazy (fun _ -> new State<'value>(init) :> IConnectable),
                renderOnChange = true
            )
        ) :?> IWritable<'value>

..
```

### Effects

Render only once after component is initialised / after first render
```fsharp
Component (fun ctx ->
    this.useEffect (
        handler = (fun _ ->
            ..
        ),
        triggers = [ EffectTrigger.AfterInit ]
    )
    ..
)
```

Render after every render.
```fsharp
Component (fun ctx ->
    this.useEffect (
        handler = (fun _ ->
            ..
        ),
        triggers = [ EffectTrigger.AfterRender ]
    )
    ..
)
```

Render when dependencies changed. 
```fsharp
Component (fun ctx ->
    let contacts = ctx.usePassedState ContactStore.shared.Contacts
    let selectedId = ctx.useState (None: Guid option)

    this.useEffect (
        handler = (fun _ ->
            ..
        ),
        triggers = [ 
            EffectTrigger.AfterChange selectedId.Any
            EffectTrigger.AfterChange selectedContact.Any
        ]
    )
    ..
)

@Zaid-Ajaj  @JordanMarr @sleepyfran Thoughts?

**Comment by JordanMarr** [comment]
> @Zaid-Ajaj @JordanMarr @sleepyfran Thoughts?

My thoughts are that you are on 🔥!

I like the React-y hook names.

One question about re-rendering when dependencies change:
```F#
        triggers = [ 
            EffectTrigger.AfterChange selectedId.Any
            EffectTrigger.AfterChange selectedContact.Any
        ]
```

Is it possible to pass in any value on the model to AfterChange?
For example passing an arbitrary property that lives on the useState model -- or does it always check the entire model?

**Comment by JaggerJo** [comment]
> > @Zaid-Ajaj @JordanMarr @sleepyfran Thoughts?
> 
> My thoughts are that you are on 🔥!
> 
> I like the React-y hook names.
> 
> One question about re-rendering when dependencies change:
> 
> ```fsharp
>         triggers = [ 
>             EffectTrigger.AfterChange selectedId.Any
>             EffectTrigger.AfterChange selectedContact.Any
>         ]
> ```
> 
> Is it possible to pass in any value on the model to AfterChange? 

`EffectTrigger.AfterChange` works with anything that implements the `IAnyReadable` interface.

Types that implement `Readable<'t>` (like `State<'t>`) are extended to have an `Any : IAnyReadable` property.

https://github.com/fsprojects/Avalonia.FuncUI/blob/7d089335946b1e3f9dd45c66172d7ece807f9c53/src/Avalonia.FuncUI.Components/State/State.fs#L46-L59

> For example passing an arbitrary property that lives on the useState model -- or does it always check the entire model?

Right now it does not even check if there was really a change to the value. It just subscribes to the passed dependencies. Those dependencies might have a unique value filter.

It's also possible to "focus" state and add a unique value filter to it.

https://github.com/fsprojects/Avalonia.FuncUI/blob/7d089335946b1e3f9dd45c66172d7ece807f9c53/src/Avalonia.FuncUI.Components/State/State.Functions.fs#L38-L43

```fsharp
Component.create ("Person", fun ctx ->
    let name =
        person
        |> State.readMap (fun person -> person.FullName)
        |> State.readUnique
        |> ctx.usePassedState (* the context already supports state that does not trigger a render on change. there are just no overloads for it yet.  *) 

    this.useEffect (
        handler = (fun _ ->
            ..
        ),
        triggers = [ 
            EffectTrigger.AfterChange name.Any
        ]
    )

    ..
)
```

**Comment by JaggerJo** [comment]
To test more advanced scenarios I ported an app ([FSharpGeneticAlgorithm](https://github.com/IntegerMan/FSharpGeneticAlgorithm)) from @IntegerMan to FuncUI and the new state management. 

There are still a few small bugs I need to fix, but I went pretty smooth overall. 

Screenshot 2021-12-22 at 20 02 24

**Comment by Zaid-Ajaj** [comment]
@JaggerJo This looks absolutely amazing 😍 😍 can't wait to try it out!

**Comment by JordanMarr** [comment]
Nice! 
I plan on trying out your component system by recreating my [chord parser](https://github.com/JordanMarr/FSharp.ChordParser) UI in the very near future. 
Of course I will need to contrive a reason to have more than one screen to fully test the components aspect. 🤔

**Comment by JaggerJo** [comment]
> Nice!
> I plan on trying out your component system by recreating my [chord parser](https://github.com/JordanMarr/FSharp.ChordParser) UI in the very near future.
> Of course I will need to contrive a reason to have more than one screen to fully test the components aspect. 🤔

Could be a nice app to include in the Examples if that would be ok for you.

**Comment by sleepyfran** [comment]
This is absolutely amazing, @JaggerJo, really really great work! The latest changes look amazing. What's left for this to be ready for review? Planning on taking a deep look soon, just want to know if you think it's ready or still needs some time 😄

**Comment by JaggerJo** [comment]
> This is absolutely amazing, @JaggerJo, really really great work! The latest changes look amazing. What's left for this to be ready for review? Planning on taking a deep look soon, just want to know if you think it's ready or still needs some time 😄

Thanks for the compliment 🙂

There is still a lot to work on and discuss.

## General
- [ ] G.1 Counterpart to `useRef` needs to be finished. 

Right now `RenderOnChange` is always set to `true`

https://github.com/fsprojects/Avalonia.FuncUI/blob/86e68fd76f6cd93907a39cc0388263ecf30c2915/src/Avalonia.FuncUI.Components/Context/Context.StateHook.fs#L53-L56

- [ ] G.2 API needs some polishing. 

- `ctx.usePassedStateReadOnly` is way too long IMHO. Alternatives could be `ctx.read.usePassed` - and for rw `ctx.usePassed`. 

- Convention for read only State functions / functions that return `IReadable<'value>`.

https://github.com/fsprojects/Avalonia.FuncUI/blob/86e68fd76f6cd93907a39cc0388263ecf30c2915/src/Avalonia.FuncUI.Components/State/State.Functions.fs#L36-L48

- [ ] G.3 Add more State functions to fill the gaps. 

While porting the Genetic Algorithm UI I notices that a `map` function was missing, so I quickly added one. If we release this with a lot of missing state transformation functions potential users could be frustrated quickly. So I'd like to port a few more apps and see what's still missing.  

- [ ] G.4 Code needs documentation comments and in general more comments.

## Tests
- [ ] T.1 Tests need to be fixed (currently don't compile)
- [ ] T.2 Tests regarding keyed components should be added [optional]
- [ ] T.3 Tests for adapters should be added [optional]

**Comment by sleepyfran** [comment]
Thanks for the detailed answer @JaggerJo! Do you want to divide the effort a bit? I might be able to help you with the tests and such 😃

**Comment by JaggerJo** [comment]
> Thanks for the detailed answer @JaggerJo! Do you want to divide the effort a bit?

@sleepyfran yes, happy to split the work. 

>  I might be able to help you with the tests and such 😃

Would be great! If you see anything that's not immediately clear feel free to attach a `?` comment in this PR and I'll try to add a better comment.

**Comment by JaggerJo** [comment]
Worked on a simple inspector / debug tool. Still an early prototype, but I think this has some potential.

image

image

**Comment by JaggerJo** [comment]
Inspector Prototype merged!

Had to get rid of a few features I really wanted (visual selection, sending values, better value view), but they took way too much time for something that initially was intended to be a proof of concept. 

image

**Comment by JaggerJo** [comment]
`ctx.attrs` can be used to attache attributes directly to the component control. 

This does not work because the `Dock` property is set on the TextBlock, not on the component directly. 
```fsharp
DockPanel.create [
    DockPanel.children [
        Component.create ("", fun ctx ->
            TextBlock.create [
                TextBlock.dock Dock.Top
                TextBlock.text "Should be on top!"
            ]
        )

        Component.create ("", fun ctx ->
            TextBlock.create [
                TextBlock.dock Dock.Bottom
                TextBlock.text "Should be on the bottom!"
            ]
        )
    ]
]
```

This works, as it sets the attributes on the component.
```fsharp
DockPanel.create [
    DockPanel.children [
        Component.create ("", fun ctx ->
            ctx.attrs [ Component.dock Dock.Top ]

            TextBlock.create [
                TextBlock.text "Should be on top!"
            ]
        )

        Component.create ("", fun ctx ->
            ctx.attrs [ Component.dock Dock.Top ]

            TextBlock.create [
                TextBlock.text "Should be on the bottom!"
            ]
        )
    ]
]
```

image

**Comment by JaggerJo** [comment]
It is now possible to get the underlaying reference of a view - (called an outlet in the code).

image


Screenshot 2022-01-06 at 17 35 07

**Comment by JaggerJo** [comment]
Right now all the state management / component related stuff is located in a separate project.

Thinking about merging the `Avalonia.FuncUI` and `Avalonia.FuncUI.Components` Projects. 

Screenshot 2022-01-06 at 18 02 47

After that I think this is ready for a preview/beta release.

@JordanMarr Thoughts?

**Comment by sleepyfran** [comment]
Agree with merging the components project inside of the FuncUI project. I've started working on a project using the hooks, it's really lovely to use!

Also spent a bit the other day taking a deeper look on how everything works. I imagine that when you refer to add tests for keyed components you mean to test the way the differ uses the view key to determine the diff? Is there another piece of logic that relies on the key or any other tests you can think of? 😄

**Comment by JaggerJo** [comment]
@sleepyfran 

Thanks for the patience. Currently renovating our new Apartment and about to move.

> Agree with merging the components project inside of the FuncUI project. 

👍🏼 

> I've started working on a project using the hooks, it's really lovely to use!

Nice! Were there any ruff edges? / Things we should improve?

> Also spent a bit the other day taking a deeper look on how everything works. I imagine that when you refer to add tests for keyed components you mean to test the way the differ uses the view key to determine the diff? 

Exactly.

> Is there another piece of logic that relies on the key or any other tests you can think of? 😄 

No, should only be relevant for the differ.

**Comment by sleepyfran** [comment]
@JaggerJo no worries, thanks for the response and good luck with the moving 😃 

> Nice! Were there any ruff edges? / Things we should improve?

I haven't really got past a basic UI app with a couple of screens so I might discover more stuff later, but the only thing I'm missing is a `Map` function in the `useState` hook similar to how React accepts a function instead of a value when calling `setState`. It's not really that important since you can always use the current state value, but it'd make mapping operations a bit easier. Aside from that I'd say it's been a really pleasurable experience and having that many examples from the beginning gave me a really good starting point, so props for that!

--------

On a side note I've started the process of fixing/adding tests in the [universal-hooks-tests branch](https://github.com/fsprojects/Avalonia.FuncUI/commits/universal-hooks-tests), started by solving the `Outlet` not being specified there. 

I've only encountered one problem that I'm not entirely sure how to solve: ever since the introduction of the `Outlet` property we can't really do `Assert.Equal` anymore, since `Outlet` is a function and equality checking is not supported. So not really sure how to process on equality comparing in these cases: should we ignore the actual value of the outlet and just check whether the Option Some/None matches or can you think of another way of doing it? Since the function also returns `unit` there's really not that much we can do with it.

As you can see in the branch I've basically implemented the custom equality attr in the ViewDelta, which is not the prettiest thing but it's the only way I can think of doing so. If you have any other ideas please let me know 😃 

Will update the thread once I finish the rest of the tests.

**Comment by JordanMarr** [comment]
I started a sample component UI today to better understand how it works. 

My current dilemma is that I created some state using `useState`, but when I try to set the state, it fails because it is not writable. Is there a special way I need to create It to make it writable?

**Comment by JaggerJo** [comment]
> I started a sample component UI today to better understand how it works.
> 
> My current dilemma is that I created some state using `useState`, but when I try to set the state, it fails because it is not writable. Is there a special way I need to create It to make it writable?

@JordanMarr can you give me an example?

**Comment by JaggerJo** [comment]
> @JaggerJo no worries, thanks for the response and good luck with the moving 😃
> 
> > Nice! Were there any ruff edges? / Things we should improve?
> 
> I haven't really got past a basic UI app with a couple of screens so I might discover more stuff later, but the only thing I'm missing is a `Map` function in the `useState` hook similar to how React accepts a function instead of a value when calling `setState`. It's not really that important since you can always use the current state value, but it'd make mapping operations a bit easier. Aside from that I'd say it's been a really pleasurable experience and having that many examples from the beginning gave me a really good starting point, so props for that!

@sleepyfran there currently are 2 mapping adapters, one for read and one for read + write. The write one has no convenient pipe-able function yet (was not sure about the signature, think we could do cool things with quotations..). 

image

Should be simple to add tho.

> On a side note I've started the process of fixing/adding tests in the [universal-hooks-tests branch](https://github.com/fsprojects/Avalonia.FuncUI/commits/universal-hooks-tests), started by solving the `Outlet` not being specified there.
> 
> I've only encountered one problem that I'm not entirely sure how to solve: ever since the introduction of the `Outlet` property we can't really do `Assert.Equal` anymore, since `Outlet` is a function and equality checking is not supported. So not really sure how to process on equality comparing in these cases: should we ignore the actual value of the outlet and just check whether the Option Some/None matches or can you think of another way of doing it? Since the function also returns `unit` there's really not that much we can do with it.
> 
> As you can see in the branch I've basically implemented the custom equality attr in the ViewDelta, which is not the prettiest thing but it's the only way I can think of doing so. If you have any other ideas please let me know 😃
> 
> Will update the thread once I finish the rest of the tests.

Ahh, yeah. I think we can ignore the `Outlet` field entirely.

**Comment by JordanMarr** [comment]
> t dilemma is that I created some state using `useState`, but when I try to set the state, it fails because it is not writable. Is there a special way I need to create It to make it

Never mind - it was a dumb mistake. 
I failed to notice that `useState` returns an `IWritable<'T>` because I am so used to it returning `'T` in React and Feliz.
The fix was to call `model.Current`. Somehow I missed that detail when looking at your samples.

**Comment by JordanMarr** [comment]
PR  #182 adds a ChordParser sample UI (based on my [FParsec project](https://github.com/JordanMarr/FSharp.ChordParser)).
I also added an experiment `useElmish` hook that wraps `useState` that was inspired by Feliz!

- One of the things I really like about Feliz is the ability to use `useState` for simple components or `useElmish` for more complex components.
- The `useElmish` hook also allows you to refer to `model` and `dispatch` directly in your component, so no need to call `model.Current`
- There are currently two overloads of `useElmish`: one that takes an `IWritable<'Model>' and one that takes a `'Model`.

What do you think? Should I continue work on it?

**Comment by JaggerJo** [comment]
> PR #182 adds a ChordParser sample UI (based on my [FParsec project](https://github.com/JordanMarr/FSharp.ChordParser)). I also added an experimental `useElmish` hook that wraps `useState` (inspired by Feliz)!
> 
> * One of the things I really like about Feliz is the ability to use `useState` for simple components or `useElmish` for more complex components.
> * The `useElmish` hook also allows you to refer to `model` and `dispatch` directly in your component, so no need to call `model.Current`
> * There are currently two overloads of `useElmish`: one that takes an `IWritable<'Model>' and one that takes a `'Model`.
> 
> What do you think? Should I continue work on it?

Like it, but think we should move it to the FuncUI.Elmish Project.

**Comment by sleepyfran** [comment]
I've also finished the tests for the keyed views in #183. @JaggerJo let me know if you miss anything 😃

**Comment by JordanMarr** [comment]
It would be nice if `useState` behaved the same way as it does in React and Feliz by returning a tuple.

Ex:
```F#
let filter, setFilter = useState filter
```
This is nice because it makes your UI code easier to write by eliminating the extra calls to `.Set` and `.Current`.

However, I realize that `IWritable<'value>` is fundamental for passing stuff between components.

Maybe there could be a different `useState` overload that works like classic React / Feliz?

Alternatively, maybe there could be a simple built-in function to easily convert an `IWritable<'value>` to a tuple:

```F#
/// Needs a better name that this...
let split (writable: IWritable<_>) = 
        writable.Current, writable.Set
```

### Usage
```F#
let enabled, setEnabled = ctx.useState true |> split
```

**Comment by JaggerJo** [comment]
> I've also finished the tests for the keyed views in #183. @JaggerJo let me know if you miss anything 😃

Nice!

**Comment by JaggerJo** [comment]
> It would be nice if `useState` behaved the same way as it does in React and Feliz by returning a tuple.
> 
> Ex:
> 
> ```fsharp
> let filter, setFilter = useState filter
> ```
> 
> This is nice because it makes your UI code easier to write by eliminating the extra calls to `.Set` and `.Current`.
> 
> However, I realize that `IWritable<'value>` is fundamental for passing stuff between components.
> 
> Maybe there could be a different `useState` overload that works like classic React / Feliz?
> 
> Alternatively, maybe there could be a simple built-in function to easily convert an `IWritable<'value>` to a tuple:
> 
> ```fsharp
> /// Needs a better name that this...
> let split (writable: IWritable<_>) = 
>         writable.Current, writable.Set
> ```
> 
> ### Usage
> ```fsharp
> let enabled, setEnabled = ctx.useState true |> split
> ```

I personally never liked the way react does this and would not include a function that seperates current and set right now. 

It can easily be added tho by people not liking the current model - but if you ever need to pass that state down you need to reconstruct a `Readable` somehow, making the events work, ... so I would not encourage doing this by including it in the lib.

**Review by sleepyfran — APPROVED** [review]

**Comment by JaggerJo** [comment]
Super busy right now unfortunately. 

This is ready to merge after combining the `FuncUI` and `FuncUI.Components` projects. Would someone take over that part?

**Comment by sleepyfran** [comment]
> Super busy right now unfortunately.
> 
> This is ready to merge after combining the `FuncUI` and `FuncUI.Components` projects. Would someone take over that part?

Awesome to hear it's ready! I'll take care of combining both modules, solving all the conflicts and merging tomorrow 😃

**Comment by JaggerJo** [comment]
> > Super busy right now unfortunately.
> > This is ready to merge after combining the `FuncUI` and `FuncUI.Components` projects. Would someone take over that part?
> 
> Awesome to hear it's ready! I'll take care of combining both modules, solving all the conflicts and merging tomorrow 😃

IIRC there is already a namespace + folder in the `FuncUI` project - but it's meaning is a bit different. Might be good to change that as I see the could lead to confusion.

**Comment by JaggerJo** [comment]
@sleepyfran let me know if you need help!

**Comment by sleepyfran** [comment]
> @sleepyfran let me know if you need help!

Thanks :)

I actually haven't had that much time, but I started the process the other day on the branch [universal-hooks-moving](https://github.com/fsprojects/Avalonia.FuncUI/tree/universal-hooks-moving), having some compilation issues that for some reason didn't arise when the code was in a separate project, but as I said I haven't had much time to look into it, hopefully I can finish the work soon.

How should we name the code that is currently sitting in the components folder, though? It's just the Hosts, Lazy and DataTemplateView, so not sure what the correct naming for that one should be.

**Comment by JaggerJo** [comment]
> > @sleepyfran let me know if you need help!
> 
> Thanks :)
> 
> I actually haven't had that much time, but I started the process the other day on the branch [universal-hooks-moving](https://github.com/fsprojects/Avalonia.FuncUI/tree/universal-hooks-moving), having some compilation issues that for some reason didn't arise when the code was in a separate project, but as I said I haven't had much time to look into it, hopefully I can finish the work soon.
> 
> How should we name the code that is currently sitting in the components folder, though? It's just the Hosts, Lazy and DataTemplateView, so not sure what the correct naming for that one should be.

No worries, can relate. Time is flying by right now.

Regarding the stuff currently in the components folder:

- Move hosts & data template views to top-level namespace
- Delete Lazy (components should be used instead)

**Comment by sleepyfran** [comment]
> > > @sleepyfran let me know if you need help!
> > 
> > 
> > Thanks :)
> > I actually haven't had that much time, but I started the process the other day on the branch [universal-hooks-moving](https://github.com/fsprojects/Avalonia.FuncUI/tree/universal-hooks-moving), having some compilation issues that for some reason didn't arise when the code was in a separate project, but as I said I haven't had much time to look into it, hopefully I can finish the work soon.
> > How should we name the code that is currently sitting in the components folder, though? It's just the Hosts, Lazy and DataTemplateView, so not sure what the correct naming for that one should be.
> 
> No worries, can relate. Time is flying by right now.
> 
> Regarding the stuff currently in the components folder:
> 
>     * Move hosts & data template views to top-level namespace
> 
>     * Delete Lazy (components should be used instead)

Thanks a lot! I finished the whole thing and published the PR in #188, take a look whenever you get some time, but this should be ready to roll! 😃

**Comment by sleepyfran** [comment]
It looks like there were still some conflicts even after merging my branch, should be good now. Also bumped the dotnet version in the Actions so that they don't fail while trying to compile the Chord Parser project 😃

**Comment by sleepyfran** [comment]
We're live 🚀 

Thanks @JaggerJo for the amazing work 😃

**Comment by JaggerJo** [comment]
@sleepyfran nice! @JordanMarr 

How do we go about releasing this? I guess we should also change the package name to not include 'Jaggerjo'.

**Comment by JaggerJo** [comment]
@sleepyfran nice! @JordanMarr 

How do we go about releasing this? I guess we should also change the package name to not include 'Jaggerjo'.

**Comment by sleepyfran** [comment]
> @sleepyfran nice! @JordanMarr
> 
> How do we go about releasing this? I guess we should also change the package name to not include 'Jaggerjo'.

Agree, we were having that conversation as part of #173. My favorite option was `FSharp.Avalonia` as @JordanMarr suggested, but I'm up for whatever other naming scheme. I've never published anything on NuGet, is it enough with renaming the Product/PackageId fields in the fsproj file or is there any other change required in NuGet itself?

**Comment by sleepyfran** [comment]
Also created #189 to track the steps needed for before/after the release of the new version so that the docs and other resources are in good health, feel free to suggest any other steps to be added 😃

**Pull request merged**